binary_search_tree/tree.py

```diff
@@ -8,41 +8,100 @@ def __init__(self, key, val = None):
         self.left = None
         self.right = None
         
-
+    def dict(self):
+        return {"key": self.key, "value": self.value}
 
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+
+    # Time Complexity: O(log n) if balanced
+    # Space Complexity: O(log n)
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value) 
+    
+    def add_helper(self, current_node, key, value):
+        if current_node == None:
+            return TreeNode(key, value) 
+        if key <= current_node.key:
+            current_node.left = self.add_helper(current_node.left, key, value)
+        else:
+            current_node.right = self.add_helper(current_node.right, key, value)
+        return current_node
+
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n) if balanced 
+    # Space Complexity: O(1)
     def find(self, key):
```

Comment:
👍 , but the space complexity is O(log n) if the tree is balanced and O(n) if it's not because of the recursive call stack.